File: src/main/java/uk/co/lukestevens/hibernate/JPAController.java
package uk.co.lukestevens.hibernate;

import org.hibernate.Session;
import org.hibernate.SessionFactory;
import org.hibernate.Transaction;
import uk.co.lukestevens.utils.Wrapper;

import java.io.IOException;

public interface JPAController {

    public <T> T useSession(ReturningOperation<T, Session> operation) throws IOException;

    public void useSession(VoidOperation<Session> operation) throws IOException;

    public <T> T useTransaction(ReturningOperation<T, Session> operation) throws IOException;

    public void useTransaction(VoidOperation<Session> operation) throws IOException;

    public Session getSession();

    /**
     * @return The SessionFactory instance managed by this controller,
     */
    public SessionFactory getFactory();

}

File: src/main/kotlin/uk/co/lukestevens/hibernate/JPAController.kt
package uk.co.lukestevens.hibernate

import org.hibernate.Session
import org.hibernate.SessionFactory

interface JPAController {

    fun <T> useSession(operation: (Session) -> T): T
    fun useSession(operation: (Session) -> Unit)
    fun <T> useTransaction(operation: (Session) -> T): T
    fun useTransaction(operation: (Session) -> Unit)
    fun <T> getDao(type: Class<T>): Dao<T>
    fun getSession(): Session

    /**
     * @return The SessionFactory instance managed by this controller,
     */
    val factory: SessionFactory
}